packages/beacon-node/src/chain/blocks/blockInput/blockInput.ts

```diff
@@ -0,0 +1,875 @@
+
+export function createPromise<T>(): PromiseParts<T> {
+  let resolve!: (value: T) => void;
+  let reject!: (e: Error) => void;
+  const promise = new Promise<T>((_resolve, _reject) => {
+    resolve = _resolve;
+    reject = _reject;
+  });
+  return {
+    promise,
+    resolve,
+    reject,
+  };
+}
+
+type BlockInputState<F extends ForkName> =
+  | {
+      blockStatus: BlockStatus.MissingBlock;
+      daStatus: DAStatus.IncompleteData;
+    }
+  | {
+      blockStatus: BlockStatus.MissingBlock;
+      daStatus: DAStatus.CompleteData;
+    }
+  | {
+      blockStatus: BlockStatus.HasBlock;
+      daStatus: DAStatus.IncompleteData;
+      block: SignedBeaconBlock<F>;
+      source: SourceMeta;
+    }
+  | {
+      blockStatus: BlockStatus.HasBlock;
+      daStatus: DAStatus.CompleteData;
+      block: SignedBeaconBlock<F>;
+      source: SourceMeta;
+      timeCompleteSec: number;
+    };
+
+abstract class AbstractBlockInput<F extends ForkName = ForkName, TData extends DAData = DAData>
+  implements IBlockInput<F, TData>
+{
+  abstract type: DAType;
+  daRequirement: DARequirement;
+  timeBeginSec: number;
+
+  forkName: ForkName;
+  slot: Slot;
+  blockRootHex: string;
+  parentRootHex: string;
+
+  abstract state: BlockInputState<F>;
+
+  protected blockPromise = createPromise<SignedBeaconBlock<F>>();
+  protected dataPromise = createPromise<TData>();
+  protected bothPromise = createPromise<this>();
+
+  constructor(init: BlockInputInit) {
+    this.daRequirement = init.daRequirement;
+    this.timeBeginSec = init.timeBeginSec;
+    this.forkName = init.forkName;
+    this.slot = init.slot;
+    this.blockRootHex = init.blockRootHex;
+    this.parentRootHex = init.parentRootHex;
+  }
+
+  abstract addBlock(props: AddBlock<F>): void;
+
+  hasBlock(): boolean {
+    return this.state.blockStatus === BlockStatus.HasBlock;
+  }
+
+  getBlock(): SignedBeaconBlock<F> {
+    if (this.state.blockStatus !== BlockStatus.HasBlock) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISSING_BLOCK,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot getBlock from BlockInput without a block"
+      );
+    }
+    return this.state.block;
+  }
+
+  getBlockSource(): SourceMeta {
+    if (this.state.blockStatus !== BlockStatus.HasBlock) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISSING_BLOCK,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot getBlockSource from BlockInput without a block"
+      );
+    }
+    return this.state.source;
+  }
+
+  hasData(): boolean {
+    return this.state.daStatus === DAStatus.CompleteData;
+  }
+
+  hasBlockAndData(): boolean {
+    return this.state.blockStatus === BlockStatus.HasBlock && this.state.daStatus === DAStatus.CompleteData;
+  }
+
+  getLogMeta(): LogMetaBasic {
+    return {
+      blockRoot: prettyBytes(this.blockRootHex),
```

Comment:
This pattern with `prettyBytes` gets used in a lot of places.  Ive also found that its helpful externally for error messages and log props by consumers.  We should make a getter for this.

```ts
get prettyBlockRootHex(): string {
  return prettyBytes(this.blockRootHex)
}
```

```suggestion
      blockRoot: this.prettyBlockRootHex,
```

Reply:
we agreed that this is for logging and consumers should just use `getLogMeta` for consistency in logging